packages/commonwealth/server/controllers/server_groups_methods/refresh_membership.ts

```diff
@@ -13,29 +23,97 @@ export type RefreshMembershipResult = {
 ): Promise<RefreshMembershipResult> {
-  /*
-    TODO: Check membership status of address for all groups for all topics within the chain,
-    or optionally for the single specified topic
-      - if membership missing or stale => recompute, save and return membership
-      - else if membership fresh => return membership
-
-      Membership model:
-      {
-        group_id: number
-        address_id: number
-        allowed: boolean
-        reject_reason?: string
-        last_checked: Date
+  // get all groups across the chain topics
+  const chainTopics = await this.models.Topic.findAll({
+    where: {
+      chain_id: chain.id,
+      ...(topicId ? { id: topicId } : {}),
+    },
+  });
+  const groupIds = uniq(flatten(chainTopics.map(({ group_ids }) => group_ids)));
+  const groups = await this.models.Group.findAll({
+    where: {
+      id: { [Op.in]: groupIds },
+    },
+  });
+
+  // update membership for each group
+  const updatedMemberships = await Promise.all(
+    groups.map(async (group) => {
+      const [membership, created] = await this.models.Membership.findOrCreate({
+        where: {
+          group_id: group.id,
+          address_id: address.id,
+        },
+        defaults: {
+          group_id: group.id,
+          address_id: address.id,
+          reject_reason: null,
+          last_checked: Sequelize.literal('CURRENT_TIMESTAMP') as any,
+        },
+      });
+      membership.Group = group;
+
+      if (!created) {
+        const expiresAt = moment(membership.last_checked).add(
+          MEMBERSHIP_TTL_SECONDS,
+          'seconds'
+        );
+        if (moment().isBefore(expiresAt)) {
+          // already exists and is fresh, don't recompute
+          return membership;
+        }
       }
-  */
-  return {
-    topicId: 1,
-    allowed: true,
-  };
+
+      // is newly created or stale, recompute
+      return recomputeMembership(membership, address, this.tokenBalanceCache);
+    })
+  );
+
+  // transform memberships to result shape
+  const results = updatedMemberships.map((membership) => {
+    const group = chainTopics.find((topic) =>
+      topic.group_ids.includes(membership.group_id)
+    );
+    return {
+      topicId: group.id,
+      allowed: !membership.reject_reason,
+      rejectReason: membership.reject_reason,
+    };
+  });
+
+  return results;
+}
+
+/**
+ * recomputeMembership checks the membership against the requirements,
+ * updates the membership status and returns it
+ * @param membership The membership to recompute
+ * @param address The user address
+ * @returns MembershipInstance
+ */
+async function recomputeMembership(
+  membership: MembershipInstance,
+  address: AddressInstance,
+  tokenBalanceCache: TokenBalanceCache
+): Promise<MembershipInstance> {
+  if (!membership.Group) {
+    throw new ServerError('membership Group is not populated');
+  }
+  const { requirements } = membership.Group;
+  const { isValid, messages } = validateGroupMembership(
```

Comment:
This is going to be async function in upcoming changes https://github.com/hicommonwealth/commonwealth/pull/5005
I think it makes most sense for merge order considerations to add and `await` here and update `validateGroupMembership` to return a promise - I'll comment there with more info